source/_posts/2017-11-15-code-generation-with-sourcery.md

```diff
@@ -0,0 +1,133 @@
+        try json.set(JSONKeys.companyId.rawValue, companyId)
+        return json
+    }
+}
+
+extension AppUser: NodeRepresentable {
+    public func makeNode(in context: Context?) throws -> Node {
+        var node = Node([:])
+        try node.set(idKey, id)
+        try node.set(Keys.name, name)
+        try node.set(Keys.email, email)
+        try node.set(Keys.companyId, companyId)
+        try node.set(AppUser.createdAtKey, createdAt)
+        try node.set(AppUser.updatedAtKey, updatedAt)
+        try node.set("company", company.get())
+        return node
+    }
+}
+
+extension AppUser: Preparation {
+    public static func prepare(_ database: Database) throws {
+        try database.create(self) { user in
+            user.id()
+            user.string(Keys.email, unique: true)
+            user.string(Keys.name, optional: true)
+            user.string(Keys.password)
+            user.string(Keys.companyId)
+        }
+        
+        
+        try database.index("email", for: AppUser.self)
+        try database.index(Keys.companyId, for: AppUser.self)
+    }
+    
+    public static func revert(_ database: Database) throws {
+        try database.delete(self)
+    }
+}
+
+extension AppUser {
+    public func makeRow() throws -> Row {
+        var row = Row()
+        try row.set(Keys.email, email)
+        try row.set(Keys.name, name)
+        try row.set(Keys.password, password)
+        try row.set(Keys.companyId, companyId)
+        return row
+    }
+}
+```
+
+As you can see, it's quite a bit of code. Sadly, it doesn't get much better from here. The more Vapor features you opt-in to the more boilerplate you're going to have to write. At Nodes, we noticed that most of our day was spent writing boilerplate and we began to search for solutions. After doing some research, we ended up going with [Sourcery](https://github.com/krzysztofzablocki/Sourcery).
+
+## Sourcery
+Taken from the GitHub repository:
+> Sourcery is a code generator for Swift language, built on top of Apple's own SourceKit. It extends the language abstractions to allow you to generate boilerplate code automatically.
+
+Just give Sourcery a few templates using the [Stencil](https://github.com/kylef/Stencil) template language and it'll solve all of your problems.
+
+Armed with this amazing tool, we built a collection of templates that eliminate virtually all of Vapor's boilerplate. For example, we can replace all of the previous code with the following:
+```swift
+final class AppUser: Model, Timestampable, SoftDeletable {
+    public var email: String
+    public var name: String?
+    public var password: String
+    public var companyId: Identifier
+}
+```
+
+and then run `sourcery` in the project directory and it will generate all of the initializers; `Node`, `JSON` and `Row` representable conformance and the database preparations. You can learn more about our templates [here](https://github.com/nodes-vapor/sourcery-templates).
```

Comment:
Again agree with the TODO. Think it makes sense to cover stuff like:
- Emphasize the idea between using Sourcery once to scaffold or using Sourcery all the time to generate the code (opt in and opt out)
- The upsides/downsides to opting in and opting out (e.g. one touching a generated file, how we deal with the files in the project structure)
- The idea of handling migrations as well (update it once: in the template - run sourcery on all the projects)

Reply:
Generally you can probably take some bits and pieces from the template readme and use here as well